### DISCUSSION TITLE
Design Question: Column does not have a public constructor?

### DISCUSSION THREAD

**Discussion by unficyp** [vrogier/ocilib#330] [General]
Hi,
as the title says - is there a reason why there is no public constructor for the column class?
Reason: i try to put the column into a (Qt) QVariant and read the Column back in another part of the program (for this i need to declare a Metatype with Q_DECLARE_METATYPE and for this a default /copy constructor are needed (https://doc.qt.io/qt-6/qmetatype.html#Q_DECLARE_METATYPE))

Background:
I build some lightweight IDE - this contains a datagrid for the output of queries. If someone clicks on e.g. a VARCHAR2 column, i open a hexeditor to display the contents. This works for data types like number, varchar2 because i put the content of the columns in my own (custom-)recordset during the loop. For CLOB columns, i don't have the content in the recordset (does not make any sense to fetch all *LOB column content), so i need a way to read the content on demand. A way to put 

`ocilib::Column col = rs.GetColumn(col);`

"col" "somewhere" for a possible later processing.

I don't know if that makes any sense...
(and my c++ skills are not enough yet to figure this out)

regards

**Comment by vrogier** [comment]
Hi,

A Column object only exists in the context of the component that returned it (resultset, typeinfo) as it holds a transient handle that is owned elsewhere.
That's why it does not have a default constructor (like most of ocilib C++ api objects).
A column object is just an accessor.
Either you retrieve values from the resultset (but as you said, it does not make sense in your case) either you choose another strategy.
By default resultsets are fetch forward only and cannot be fetched backward. Thus once, you've fetched the content to fill your grid, you can't go back and then need to reissue on the fly a new select to query that particular row value (possibly by rowid).
Or you can alternatively use scrollable cursors that can be freely fetch in both direction at anytime and you can even jump to specific rows. But they use more resources !

Regards,

Vincent

**Comment by unficyp** [comment]
hi,
thanks for your answer - i was already afraid that this would be more difficult.. :/
I guess the ROWID approach does not (always) work, because:
- i have to rewrite the query entered by the user to add the rowid column (or do the client libs fetch the rowid transparently? - not that i know of)
- impossible when the output comes from (nested) views or aggregates.

so i have to go the scrollable way or drop that feature.

any chance for a feature-request ? :p

thanks&regards,
gerald

**Comment by vrogier** [comment]
I might not completely understand your issue.

Adding a default constructor to Column object would not help at all for your delayed data retrieval issue. 
A column object does not have access to the row/col data content. It is just an accessor on metadata on the column type.

You issue is about 1) storing metadata on the column type or 2) display the content of a given column row later on ?

Vincent

**Comment by unficyp** [comment]
Uhm,
i see my mistake now, you are correct, column won't help at all - my issue is 2)

My Test App looks like this:
![image](https://user-images.githubusercontent.com/1231087/218960079-8fb7e1b4-e1bd-474e-a530-95bf52dbf3d9.png)

That may be the output of an user query, the user double clicks on either a NUMBER, CHAR or VARCHAR row - i can display the content in a popup because i have this content already in the datagrid. But for *LOB columns i need to fetch the content at the moment the user clicks on the column (in the example row5, column content_clob). So iguess i must use scrollable cursors to do a seek back to row 5 like you mentioned in your 1st reply...correct ?

gerald

**Reply by unficyp** [reply]
to answer my own question: yes :)
Got this working now - thanks !

**Comment by vrogier** [comment]
Good :)

Vincent
